source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRViewController.mm

```diff
@@ -149,4 +170,212 @@ - (void)render
+{
+    for(auto &elem : body)
+    {
+        switch (elem->GetElementType())
+        {
+            case CardElementType::TextBlock:
+            {
+                /// tag a base card element with unique key
+                [self tagBaseCardElement:elem];
+                /// dispatch to concurrent queue
+                std::shared_ptr<TextBlock> txtElem = std::dynamic_pointer_cast<TextBlock>(elem);
+                dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0),
+                    ^{
+                        // MarkDownParser transforms text with MarkDown to a html string
+                        std::shared_ptr<MarkDownParser> markDownParser = std::make_shared<MarkDownParser>(txtElem->GetText().c_str());
+                        NSString *parsedString = [NSString stringWithCString:markDownParser->TransformToHtml().c_str() encoding:NSUTF8StringEncoding];
+
+                        // Font and text size are applied as CSS style by appending it to the html string -- font is hard coded for now
+                        parsedString = [parsedString stringByAppendingString:[NSString stringWithFormat:@"<style>body{font-family: '%@'; font-size:%dpx;}</style>",
+                                                                              @"verdana", [ACRTextBlockRenderer getTextBlockTextSize:txtElem->GetTextSize() withHostConfig:_hostConfig]]];
+                        // Convert html string to NSMutableAttributedString, NSAttributedString knows how to apply html tags
+                        NSData *htmlData = [parsedString dataUsingEncoding:NSUTF16StringEncoding];
+                        NSDictionary *options = @{NSDocumentTypeDocumentAttribute: NSHTMLTextDocumentType};
+                        // Initializing NSMutableAttributedString for HTML rendering is very slow
+                        NSMutableAttributedString *content = [[NSMutableAttributedString alloc] initWithData:htmlData options:options documentAttributes:nil error:nil];
+
+                         dispatch_async(dispatch_get_main_queue(),
+                             ^{ __block UILabel *lab = nil; // generate key for text map from TextBlock element's id
+                                  NSString *key = [NSString stringWithCString:txtElem->GetId().c_str() encoding:[NSString defaultCStringEncoding]];
+                                  // syncronize access to text map
+                                  dispatch_sync(_serial_text_queue,
+                                      ^{
+                                           // UILabel is not ready, cashe UILabel
+                                           if(!_textMap[key]) {
+                                               _textMap[key] = content;
+                                           } // UILable is ready, get label
+                                           else {
+                                               lab = _textMap[key];
+                                           }
```

Comment:
I think it's confusing the way this map is used, mapping keys sometimes to strings and sometimes to UI elements. I also think the business of temporarily modifying the card's ID (to disambiguate between cards with the same ID?) seems fishy. But I don't understand this code base well enough to suggest any alternatives, so I'm letting those things pass.

Reply:
thanks, the code is initializing NSAttributedString's html rendering in the background while UI elements tree is being rendered. Either two can complete in any order, so I used NSDictionary that can get hold of any object types. 
Because we are using unique ID, there is no ambiguity of which type of the two will be stored in the dictionary. 

I wanted to use address of base elements as key, but it was unstable, and they were keep changing. so i had to create new id, and i piggyback on the existing ID. elements are not accessed again until UI elements complete rendering, so this shouldn't affect any pre-existing assumptions 

---
In reply to: [168017326](https://github.com/Microsoft/AdaptiveCards/pull/1124#discussion_r168017326) [](ancestors = 168017326)